grafana/grafana
# PanelLibrary: adding library panels to Dashboard Api

**What this PR does / why we need it**:
This PR is a first exploratory POC of how loading of dashboards could work with Panel Library. I've done some workarounds I'm not particularly happy with, but that's because I don't know better. I'll explain them in comments below.

This draft only works for loading dashboards at the moment, saving is not implemented yet.

My thinking is that a library panel model json will have an extra property if it's connected to a dashboard like so:
```json
...
"libraryPanel": {
  "name": "<some unique name>",
  "uid": "<someuid>"
},
...
```

When loading a dashboard we will loop through all the panels in the dashboard json and if we find the `libraryPanel` property we try to match it with the connected library panels from the database. If the match is ok, then we replace the panel in the dashboard json with the library panel from the database and add back any dashboard specific properties (id, gridPos at the moment).

When storing a dashboard ~~(not implemented yet)~~ we ~~would need to~~ do the reverse and loop through all panels and find library panels. For each library panel we would just keep the `libraryPanel` property and any dashboard specific properties  (id, gridPos at the moment).

When deleting a dashboard ~~(not implemented yet)~~ we ~~would need to~~ remove any existing connections.

**To test this PR**
1. Add feature toggle
```
   [feature_toggles]
    # enable features, separated by spaces
    enable = panelLibrary
```
2. Run SQL script
```sql
DELETE FROM library_panel WHERE ID > 0;
DELETE FROM library_panel_dashboard WHERE ID > 0;

INSERT INTO library_panel (id, org_id, folder_id, uid, name, model, created, created_by, updated, updated_by) VALUES (1,1,2, 'g1sNpCaMz','React Gauge','{"cacheTimeout":null,"datasource":null,"fieldConfig":{"defaults":{"color":{"mode":"thresholds"},"custom":{},"links":[{"targetBlank":true,"title":"Go to drilldown","url":"/d/O6GmNPvWk/dashboard-tests-nested-template-variables-drilldown?orgId=1&${__all_variables}&${__url_time_range}"}],"mappings":[],"max":100,"min":0,"nullValueMode":"connected","thresholds":{"mode":"absolute","steps":[{"color":"green","value":null},{"color":"red","value":80}]},"unit":"none"},"overrides":[]},"gridPos":{"h":9,"w":4,"x":16,"y":0},"id":8,"links":[],"options":{"orientation":"horizontal","reduceOptions":{"calcs":["mean"],"fields":"","values":false},"showThresholdLabels":false,"showThresholdMarkers":true,"text":{}},"pluginVersion":"7.4.0-pre","targets":[{"refId":"A","scenarioId":"random_walk"}],"timeFrom":null,"timeShift":null,"title":"React gauge datalink","type":"gauge"}','2021-01-14 10:12:22',1,'2021-01-14 10:12:22',1);
INSERT INTO library_panel (id, org_id, folder_id, uid, name, model, created, created_by, updated, updated_by) VALUES (2,1,2, 'MAnX2ifMk','React Table','{"datasource":"gdev-testdata","fieldConfig":{"defaults":{"color":{"mode":"thresholds"},"custom":{"align":"center","displayMode":"color-background","filterable":false},"mappings":[],"thresholds":{"mode":"percentage","steps":[{"color":"green","value":null},{"color":"blue","value":20},{"color":"orange","value":60},{"color":"red","value":70}]},"unit":"degree"},"overrides":[{"matcher":{"id":"byName","options":"Max"},"properties":[{"id":"custom.width","value":84}]},{"matcher":{"id":"byName","options":"Last"},"properties":[{"id":"custom.width","value":78}]},{"matcher":{"id":"byName","options":"Mean"},"properties":[{"id":"custom.width","value":74}]},{"matcher":{"id":"byName","options":"Field"},"properties":[{"id":"custom.align","value":"left"}]}]},"gridPos":{"h":16,"w":7,"x":0,"y":1},"id":4,"options":{"showHeader":true,"sortBy":[{"desc":true,"displayName":"Last"}]},"pluginVersion":"7.4.0-pre","targets":[{"refId":"A","scenarioId":"random_walk","seriesCount":15,"stringInput":""}],"timeFrom":null,"timeShift":null,"title":"Colored background","transformations":[{"id":"reduce","options":{"reducers":["max","mean","last"]}}],"type":"table"}','2021-01-14 10:12:22',1,'2021-01-14 10:12:22',1);

INSERT INTO library_panel_dashboard (id, librarypanel_id, dashboard_id, created, created_by)
SELECT 1,1,id, '2021-01-14 05:40:52',1 FROM dashboard WHERE title = 'Panel - Panel Library';
INSERT INTO library_panel_dashboard (id, librarypanel_id, dashboard_id, created, created_by)
SELECT 2,2,id, '2021-01-14 05:41:15',1 FROM dashboard WHERE title = 'Panel - Panel Library';
```

3. Navigate to [http://localhost:3000/d/imQX6j-Gz/panel-panel-library?orgId=1](http://localhost:3000/d/imQX6j-Gz/panel-panel-library?orgId=1)
4. You should see the following:
![image](https://user-images.githubusercontent.com/562238/104585903-04f85000-5665-11eb-819b-13c06a030ce6.png)
5. Use the API to update the library panels and they should be updated 🤞 

**Which issue(s) this PR fixes**:
Relates #29610

**Special notes for your reviewer**:
